Skip premature Shutoff/Unknown events

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before startup the domain is in state Shutoff with
Unknown status.
If this event is thrown to the handler, it will be
mis-interpreted and could wrongly move the vmi to
Failed state.

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  
- Fixes #
- Partially addresses #
-->
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issues/<vep_tracking_issue_number>
-->
- jira-issue: https://issues.redhat.com/browse/CNV-68835

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

fossedihelm commented:
/test pull-kubevirt-e2e-k8s-1.34-sig-compute
/test pull-kubevirt-e2e-k8s-1.34-sig-performance
/test pull-kubevirt-e2e-k8s-1.34-sig-compute-arm64
/test pull-kubevirt-unit-test-s390x

fossedihelm commented:
/test all

fossedihelm commented:
/retest

fossedihelm commented on pkg/virt-handler/vm.go:
```diff
@@ -1504,6 +1506,12 @@ func (c *VirtualMachineController) sync(key string,
+func isRecentlyCreatedDomain(domain *api.Domain) bool {
+	return domain.Status.Status == api.Shutoff && domain.Status.Reason == api.ReasonUnknown &&
+		domain.Spec.Metadata.KubeVirt.StartTimestamp != nil &&
+		time.Since(domain.Spec.Metadata.KubeVirt.StartTimestamp.Time) < 30*time.Second
```
It should not, on the contrary, it should actually make the handler more tolerant to slowness

fossedihelm committed:
virt-launcher: add StartingUp field to metadata

This will be used by the handler to be more resilient
during the startup.

Signed-off-by: fossedihelm <ffossemo@redhat.com>

fossedihelm committed:
virt-launcher: add unit tests for StartingUp field

Signed-off-by: fossedihelm <ffossemo@redhat.com>

fossedihelm committed:
virt-handler: be more resilient on Shutoff/Unknown status

In libvirt, a Domain status of Shutoff with reason Unknown
acts as a default or fallback category.
So this can appear in several scenarios:
- newly defined domains
- libvirt restart/reconnect failures
- qemu process dies
- etc...

Currently, the handler threat the Shutoff/Unknown domain
as a failure.
In high load/density scenario, it could be possible that
the launcher just need more time.

Use the new StartingUp metadata field to tolerate this.
Fail the vmi in other cases.

Signed-off-by: fossedihelm <ffossemo@redhat.com>

fossedihelm committed:
virt-handler: fail fast when the socket is gone

The socket should always be there for the entire lifecycle.
If the socket disappear it have to be due to pod crash.
In such case we can consider the domain not alive anymore.

To achieve this we will distinguish recoverable vs irrecoverable
launcher client errors.
This enables better error handling logic, particularly for scenarios
where the virt-launcher pod has been torn down.

Assisted-By: Claude <noreply@anthropic.com>
Signed-off-by: fossedihelm <ffossemo@redhat.com>

fossedihelm commented:
> > @vladikr As per offline discussion, I've moved the logic from the handler to the launcher. PTAL
> 
> Perfect! Thank you!
> 
> @fossedihelm , to be more paranoid, I would add an explicit socket check either in the sync or before that in execute. Something as we have in getLauncherClient:
> 
> ```
> launcherSocket, socketErr := cmdclient.FindSocket(vmi)
> if socketErr != nil {
>     log.Log.Reason(socketErr).Errorf("Virt-launcher command socket not found for VMI %s/%s", vmi.Namespace, vmi.Name)
>     domainExists = false
> }```
> ```

Great suggestion. Added in the last commit. Thanks @vladikr

fossedihelm committed:
notify(test): Use Receive() instead of time.After

Signed-off-by: fossedihelm <ffossemo@redhat.com>

0xFelix reviewed (approved):
/lgtm 

Thanks!

vladikr commented:
@xpivarc, it's waiting for your lgtm as I approved the PR.

xpivarc commented on pkg/virt-handler/vm.go:
```diff
@@ -1376,6 +1376,13 @@ func (c *VirtualMachineController) sync(key string,
 	}
 
+	if vmiExists {
+		if _, err := c.launcherClients.GetVerifiedLauncherClient(vmi); err != nil {
```
This will just perform ping which I don't think is enough, in the offline discussion we discussed we need to also extend this method for socket check, which would be critical and only then we should fail.

xpivarc reviewed (commented):
Thank you! I still see some issues, please see comments.

Also can we have a tracker for Libvirt to look at this spurious shutdowns? Also please open tracker for the issue we saw while debugging the problem, that is the launcher could still live...

fossedihelm committed:
Add comprehensive unit tests for launcher client verification in virt-handler:

1. In vm_test.go:
   - Test sync behavior when GetVerifiedLauncherClient fails, ensuring
     domainAlive is set to false and the domain is properly cleaned up
   - Test normal sync flow when GetVerifiedLauncherClient succeeds

2. In launcher-clients_test.go:
   - Test early return when Ping succeed (FindSocket not called)
   - Test error handling when both Ping and FindSocket fails
   - Test error handling when Ping fails and FindSocket succeeds

These tests ensure proper handling of launcher client connectivity
and socket verification, particularly for edge cases where pods are
torn down during verification.

Assisted-By: Claude <noreply@anthropic.com>

Signed-off-by: fossedihelm <ffossemo@redhat.com>

fossedihelm commented:
/retest
